### PR TITLE
🌱 Add a common patcher and adopt it in registration

### PR DESCRIPTION
**Pull request by qiujian16**
## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/common/patcher/patcher.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+// Resource is a generic wrapper around resources so we can generate patches.
+type Resource[Sp any, St any] struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              Sp `json:"spec"`
+	Status            St `json:"status,omitempty"`
+}
+
+type patchFunc[R runtime.Object, Sp any, St any] func(context.Context, R, *Resource[Sp, St], *Resource[Sp, St], ...string) error
```
Comment (zhujian7) [inline pkg/common/patcher/patcher.go:35]: I do not see this being used in other places except line 39.

File: pkg/common/patcher/patcher.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+type patcher[R runtime.Object, Sp any, St any] struct {
+	client    PatchClient[R]
+	patchFunc patchFunc[R, Sp, St]
```
Comment (zhujian7) [inline pkg/common/patcher/patcher.go:39]: why do we need this `patchFunc patchFunc[R, Sp, St]`? can it be deleted?

File: pkg/common/patcher/patcher_test.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+			validateActions: func(t *testing.T, actions []clienttesting.Action) {
+				testingcommon.AssertActions(t, actions, "patch")
+				patch := actions[0].(clienttesting.PatchAction).GetPatch()
+				managedCluster := &clusterv1.ManagedCluster{}
+				err := json.Unmarshal(patch, managedCluster)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if !equality.Semantic.DeepEqual(managedCluster.Status, newManagedClusterWithConditions(metav1.Condition{Type: "Type2"}).Status) {
+					t.Errorf("not patched correctly got %v", managedCluster.Spec)
```
Comment (zhujian7) [inline pkg/common/patcher/patcher_test.go:176]:
```suggestion
					t.Errorf("not patched correctly got %v", managedCluster.Status)
```

File: pkg/common/patcher/patcher.go (hunk 1 above)
Comment (qiujian16) [reply pkg/common/patcher/patcher.go:35]: oops we do not

**Pull request merged**